src/rpc/blockchain.cpp

```diff
@@ -1597,6 +1602,335 @@ UniValue savemempool(const JSONRPCRequest& request)
+            // add legacy P2PKH script
+            CKeyID address = pubKey.GetID();
+            CScript script = GetScriptForDestination(address);
+            if (!script.empty()) {
+                needles.insert(script);
+                temp_keystore.AddWatchOnly(script);
+            }
+            // add P2SH-P2WPKH script
+            CScript witscript = GetScriptForWitness(script);
+            CTxDestination result;
+            ExtractDestination(witscript, result);
+            CScript p2wpkh_script = GetScriptForDestination(result);
+            CScript p2sh_p2wpkh_script = GetScriptForDestination(CScriptID(p2wpkh_script));
+            if (!p2wpkh_script.empty()) {
+                needles.insert(p2wpkh_script);
+                temp_keystore.AddWatchOnly(p2wpkh_script);
+            }
+            if (!p2sh_p2wpkh_script.empty()) {
+                needles.insert(p2sh_p2wpkh_script);
+                temp_keystore.AddWatchOnly(p2sh_p2wpkh_script);
+            }
+            // add 1of1 multisig (used by Bitpay for a while)
+            script = GetScriptForMultisig(1, std::vector<CPubKey>(1, pubKey));
+            if (!script.empty()) {
+                needles.insert(script);
+                temp_keystore.AddWatchOnly(script);
+            }
+            // add legacy P2PK
+            script = GetScriptForRawPubKey(pubKey);
+            if (!script.empty()) {
+                needles.insert(script);
+                temp_keystore.AddWatchOnly(script);
+            }
+        }
+        // look for posssible addresses
+        UniValue addresses_uni = find_value(request.params[1], "addresses");
+        if (addresses_uni.isArray()) {
+            for (const UniValue& address_uni : addresses_uni.get_array().getValues()) {
+                CTxDestination dest = DecodeDestination(address_uni.get_str());
+                if (!IsValidDestination(dest)) {
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid address");
+                }
+                CScript script = GetScriptForDestination(dest);
+                if (!script.empty()) {
+                    needles.insert(script);
+                    temp_keystore.AddWatchOnly(script);
+                }
+            }
+        }
+        if (!pubkeys_uni.isArray() && !xpubs_uni.isArray() && !addresses_uni.isArray()) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid or empty publickey, addresses or xpub array");
+        }
+
+        CMutableTransaction tx;
+        // Scan the unspent transaction output set for inputs
+        UniValue unspents(UniValue::VARR);
+        std::vector<CTxOut> input_txos;
+        std::map<COutPoint, Coin> coins;
+        g_should_abort_scan = false;
+        g_scan_progress = 0;
+        int64_t count = 0;
+
+        // flush utxo state and start the scan
+        FlushStateToDisk();
+        bool res = pcoinsdbview->FindScriptPubKey(g_scan_progress, g_should_abort_scan, count, needles, coins);
```

Comment:
What prevents the state from being mutated out from under this scan?

Reply:
I think due to the time required to perform a scan, it's something that may be tolerated (although it should be mentioned in the docs). Not sure, but I guess it's the same with `gettxoutsetinfo`.

Not sure if you can scan a `CCoinsView` of a snapshot state... I guess no.
Locking cs_main would be "meh".

Reply:
Also, the rawtx (sweep) is in the same way "outdated" the moment you have received it... maybe you could argue that this is the same for `fundrawtx`

Reply:
@gmaxwell @jonasschnelli The cursor iterates over the state of the CCoinsView at the time it was created; modifying it during iteration is fine. This only works because `GetCursor` is not implemented for `CCoinsViewCache`, and is invoked directly on the `CCoinsViewDB` LevelDB wrapper.

The downside is that this requires a full flush of the database, hurting performance for all of the process (including validation).